### PR TITLE
NodeProvider has a client pool to re-use connections

### PR DESCRIPTION
**Pull request by christos-h**
## Motivation

Clients are instantiating connections multiple times to validators instead of re-using connections resulting in re-negotiating TLS and general unecessary network overhead.

## Proposal

Add a pool of clients to `NodeProvider` which can be re-used.

This decreases end-to-end operation latency by 10%-35% depending on the RTT between client and committee. 

## Test Plan

The existing tests will catch regressions. Below are some ad-hoc benchmarks:

```
RTT: 0.2ms
Transfer Operation (before): 128 ms
Transfer Operation (after): 112 ms

RTT: 50ms
Transfer Operation (before): 1450 ms
Transfer Operation (after): 997 ms
```

### CODE REVIEW

File: linera-rpc/src/node_provider.rs [hunk 1]
```diff
@@ -32,14 +39,22 @@ impl ValidatorNodeProvider for NodeProvider {
 
     fn make_node(&self, address: &str) -> anyhow::Result<Self::Node, NodeError> {
         let address = address.to_lowercase();
+        let mut pool = self.client_pool.lock().unwrap();
```
Comment (christos-h) [inline linera-rpc/src/node_provider.rs:42]: I don't think this can panic and poison the lock, that's why the unwrap is here.

**Review by afck — APPROVED** [review]

File: linera-rpc/src/node_provider.rs [hunk 2]
```diff
@@ -32,14 +39,22 @@ impl ValidatorNodeProvider for NodeProvider {
...
+        if let Some(client) = pool.get(&address) {
+            return Ok(client.clone());
```
Comment (afck) [inline linera-rpc/src/node_provider.rs:44]: Do both clients handle reconnecting if the other side dropped the connection?

Comment (ma2bd) [reply linera-rpc/src/node_provider.rs:44]: Exactly. This needs to be tested

**Comment by ma2bd** [comment]
Perhaps we can look into the approach of https://github.com/linera-io/linera-protocol/pull/2527 instead

File: linera-rpc/src/node_provider.rs [hunk 3]
```diff
@@ -10,11 +15,12 @@ use crate::{client::Client, grpc::GrpcNodeProvider};
 
 /// A general node provider which delegates node provision to the underlying
 /// node provider according to the `ValidatorPublicNetworkConfig`.
-#[derive(Copy, Clone)]
+#[derive(Clone)]
 pub struct NodeProvider {
     grpc: GrpcNodeProvider,
     #[cfg(with_simple_network)]
     simple: SimpleNodeProvider,
+    client_pool: Arc<Mutex<HashMap<String, Client>>>,
```
Comment (ma2bd) [inline linera-rpc/src/node_provider.rs:23]: Based on the experience of #2664, the `GrpcNodeProvider` should instead include a `HashMap<String, tonic::transport::Channel>`

**Review by ma2bd — CHANGES_REQUESTED** [review]

**Comment by christos-h** [comment]
Closing in favour of #2708.

**Pull request closed without merging**